File: sdk/kotlin/idenqa/src/main/kotlin/dev/idenqa/sdk/JourneyPlan.kt
package dev.idenqa.sdk

/**
 * Derives the ordered one-task-at-a-time capture plan from the immutable session
 * snapshot, the advertised capabilities, and authoritative accepted progress.
 *
 * Unknown namespaced methods fail closed: a method that is not declared by this
 * SDK and currently available on the device is never selected, and no local
 * fallback is invented. Only policy-approved fallbacks are used.
 */
internal object CapturePlanBuilder {
    data class SelectedAcquisition(val methods: List<String>, val fallback: CaptureFallbackCondition?)

    fun build(
        session: CaptureSessionDetail,
        capabilities: CaptureCapabilities,
        preferences: List<String>,
        progress: List<CaptureCompletion>,
        captureFailedTaskIds: Set<String>,
    ): CapturePlan {
        if (session.requirements.requirements.isEmpty()) throw IdenqaException.InvalidResponse
        val tasks = mutableListOf<CaptureTask>()
        for (requirement in session.requirements.requirements) {
            if (requirement.key.isEmpty() || requirement.evidenceType.isEmpty() || requirement.artefacts.isEmpty() ||
                requirement.artefacts.any { it.isEmpty() }
            ) throw IdenqaException.InvalidResponse
            val selected = selectAcquisition(
                requirement.acquisition, requirement.fallbacks, capabilities, preferences,
            )
            var requirementTasks = makeTasks(requirement, selected.methods, selected.fallback, null)
            if (selected.fallback == null) {
                requirementTasks = applyCaptureFailure(
                    requirement, requirementTasks, capabilities, preferences, captureFailedTaskIds,
                )
            }
            tasks += requirementTasks
        }
        return CapturePlan(
            verificationId = session.id,
            profileRevision = session.profileRevision,
            profileDigest = session.profileDigest,
            tasks = tasks,
            completedTaskIds = completionIds(tasks, progress),
        )
    }

    fun selectAcquisition(
        acquisition: CaptureAcquisition,
        fallbacks: List<CaptureFallback>,
        capabilities: CaptureCapabilities,
        preferences: List<String>,
    ): SelectedAcquisition {
        if (acquisition.strategy != "any_of" && acquisition.strategy != "all_of") throw IdenqaException.InvalidResponse
        val primary = evaluate(acquisition, capabilities, preferences)
        primary.methods?.let { return SelectedAcquisition(it, null) }
        for (fallback in fallbacks) {
            val condition = fallbackCondition(fallback, primary.unavailable) ?: continue
            if (condition == CaptureFallbackCondition.CAPTURE_FAILED) continue
            val candidate = evaluate(fallback.acquisition, capabilities, preferences)
            candidate.methods?.let { return SelectedAcquisition(it, condition) }
        }
        throw IdenqaException.NoCompatibleMethod
    }

    private data class Evaluation(val methods: List<String>?, val unavailable: List<CaptureFallbackCondition>)

    private fun evaluate(
        acquisition: CaptureAcquisition,
        capabilities: CaptureCapabilities,
        preferences: List<String>,
    ): Evaluation {
        val methods = acquisition.methods
        if (methods.isEmpty() || methods.toSet().size != methods.size) return Evaluation(null, emptyList())
        val supported = methods.filter { it in capabilities.declaredMethods }
        val available = methods.filter { it in capabilities.availableMethods }
        val ordered = orderedMethods(available, preferences)
        if (acquisition.strategy == "any_of" && ordered.isNotEmpty()) return Evaluation(ordered, emptyList())
        if (acquisition.strategy == "all_of" && ordered.size == methods.size) return Evaluation(ordered, emptyList())
        val unavailable = mutableListOf<CaptureFallbackCondition>()
        if (supported.size != methods.size) unavailable += CaptureFallbackCondition.METHOD_UNAVAILABLE
        if (supported.any { it !in capabilities.availableMethods }) unavailable += CaptureFallbackCondition.CAPABILITY_UNAVAILABLE
        return Evaluation(null, unavailable)
    }

    private fun orderedMethods(methods: List<String>, preferences: List<String>): List<String> {
        if (preferences.isEmpty()) return methods
        val preferred = preferences.filter { it in methods }
        if (preferred.isEmpty()) return methods
        return preferred + methods.filter { it !in preferred }
    }

    private fun fallbackCondition(
        fallback: CaptureFallback,
        unavailable: List<CaptureFallbackCondition>,
    ): CaptureFallbackCondition? = fallback.on
        .mapNotNull(CaptureFallbackCondition::fromWire)
        .firstOrNull { it != CaptureFallbackCondition.CAPTURE_FAILED && it in unavailable }

    private fun makeTasks(
        requirement: CaptureRequirementDocument,
        methods: List<String>,
        fallbackCondition: CaptureFallbackCondition?,
        artefacts: List<String>?,
    ): List<CaptureTask> {
        val values = artefacts ?: requirement.artefacts
        if (methods.size == 1 || requirement.acquisition.strategy == "any_of") {
            return values.map { task(requirement, it, methods, fallbackCondition) }
        }
        return values.flatMap { artefact ->
            methods.map { method -> task(requirement, artefact, listOf(method), fallbackCondition) }
        }
    }

    private fun applyCaptureFailure(
        requirement: CaptureRequirementDocument,
        tasks: List<CaptureTask>,
        capabilities: CaptureCapabilities,
        preferences: List<String>,
        failedTaskIds: Set<String>,
    ): List<CaptureTask> {
        if (failedTaskIds.isEmpty() || tasks.none { it.id in failedTaskIds }) return tasks
        val failedArtefacts = tasks.filter { it.id in failedTaskIds }.map { it.artefact }.toSet()
        val replacement = mutableMapOf<String, List<CaptureTask>>()
        for (fallback in requirement.fallbacks) {
            if (CaptureFallbackCondition.CAPTURE_FAILED.wire !in fallback.on) continue
            val methods = evaluate(fallback.acquisition, capabilities, preferences).methods ?: continue
            for (artefact in failedArtefacts) {
                replacement.getOrPut(artefact) { makeTasks(requirement, methods, CaptureFallbackCondition.CAPTURE_FAILED, listOf(artefact)) }
            }
        }
        if (failedArtefacts.any { it !in replacement }) throw IdenqaException.NoCompatibleMethod
        return tasks.flatMap { replacement[it.artefact] ?: listOf(it) }
    }

    private fun task(
        requirement: CaptureRequirementDocument,
        artefact: String,
        methods: List<String>,
        fallback: CaptureFallbackCondition?,
    ): CaptureTask = CaptureTask(
        id = taskId(requirement.key, artefact, methods, fallback),
        requirementKey = requirement.key,
        evidenceType = requirement.evidenceType,
        artefact = artefact,
        methodOptions = methods,
        fallbackCondition = fallback,
        requiredAssurances = requirement.requiredAssurances,
    )

    fun taskId(
        requirementKey: String,
        artefact: String,
        methods: List<String>,
        fallback: CaptureFallbackCondition?,
    ): String = listOf(requirementKey, artefact, methods.joinToString("+"), fallback?.wire ?: "").joinToString("::")

    private fun completionIds(tasks: List<CaptureTask>, progress: List<CaptureCompletion>): Set<String> {
        val completed = mutableSetOf<String>()
        for (completion in progress) {
            val matches = tasks.filter { task ->
                task.requirementKey == completion.requirementKey &&
                    task.evidenceType == completion.evidenceType &&
                    task.artefact == completion.artefact &&
                    completion.acquisitionMethod in task.methodOptions &&
                    (task.fallbackCondition?.wire ?: "") == (completion.fallbackCondition ?: "")
            }
            if (matches.size != 1) throw IdenqaException.InvalidResponse
            if (!completed.add(matches[0].id)) throw IdenqaException.InvalidResponse
        }
        return completed
    }
}

File: sdk/swift/Sources/Idenqa/JourneyPlan.swift
import Foundation

/// Derives the ordered one-task-at-a-time capture plan from the immutable session
/// snapshot, the advertised capabilities, and authoritative accepted progress.
///
/// Unknown namespaced methods fail closed: a method that is not declared by this
/// SDK and currently available on the device is never selected, and no local
/// fallback is invented. Only policy-approved fallbacks are used.
enum CapturePlanBuilder {
    static func build(
        session: CaptureSessionDocument,
        capabilities: CaptureCapabilities,
        preferences: [String],
        progress: [CaptureCompletionDocument],
        captureFailedTaskIDs: Set<String>
    ) throws -> CapturePlan {
        guard !session.requirements.requirements.isEmpty else { throw IdenqaError.invalidResponse }
        var tasks: [CaptureTask] = []
        for var requirement in session.requirements.requirements {
            if let selected = session.documentSelections?[requirement.key] {
                guard let option = requirement.documentOptions?.first(where: { $0.id == selected }) else { throw IdenqaError.invalidResponse }
                requirement.artefacts = option.artefacts
            }
            // Core canonicalises names lexically; the subject captures front first.
            requirement.artefacts.sort { left, right in
                func order(_ value: String) -> Int { value.hasSuffix(".document_front") ? 0 : value.hasSuffix(".document_back") ? 1 : 2 }
                return order(left) < order(right)
            }
            guard !requirement.key.isEmpty, !requirement.evidenceType.isEmpty, !requirement.artefacts.isEmpty,
                  !requirement.artefacts.contains(where: \.isEmpty) else {
                throw IdenqaError.invalidResponse
            }
            let selected = try selectAcquisition(
                acquisition: requirement.acquisition,
                fallbacks: requirement.fallbacks,
                capabilities: capabilities,
                preferences: preferences
            )
            var requirementTasks = makeTasks(
                requirement: requirement,
                methods: selected.methods,
                fallbackCondition: selected.fallback
            )
            if selected.fallback == nil {
                requirementTasks = try applyingCaptureFailure(
                    requirement: requirement,
                    tasks: requirementTasks,
                    capabilities: capabilities,
                    preferences: preferences,
                    failedTaskIDs: captureFailedTaskIDs
                )
            }
            tasks.append(contentsOf: requirementTasks)
        }
        let completed = try completionIDs(tasks: tasks, progress: progress)
        return CapturePlan(
            verificationID: session.id,
            profileRevision: session.profileRevision,
            profileDigest: session.profileDigest,
            tasks: tasks,
            completedTaskIDs: completed
        )
    }

    struct SelectedAcquisition {
        let methods: [String]
        let fallback: CaptureFallbackCondition?
    }

    static func selectAcquisition(
        acquisition: CaptureAcquisition,
        fallbacks: [CaptureFallback],
        capabilities: CaptureCapabilities,
        preferences: [String]
    ) throws -> SelectedAcquisition {
        let primary = evaluate(acquisition, capabilities: capabilities, preferences: preferences)
        if let methods = primary.methods { return SelectedAcquisition(methods: methods, fallback: nil) }
        for fallback in fallbacks {
            guard let condition = fallbackCondition(fallback, unavailable: primary.unavailable) else { continue }
            guard condition != .captureFailed else { continue }
            let candidate = evaluate(fallback.acquisition, capabilities: capabilities, preferences: preferences)
            if let methods = candidate.methods {
                return SelectedAcquisition(methods: methods, fallback: condition)
            }
        }
        throw IdenqaError.noCompatibleMethod
    }

    private struct Evaluation {
        let methods: [String]?
        let unavailable: [CaptureFallbackCondition]
    }

    private static func evaluate(
        _ acquisition: CaptureAcquisition,
        capabilities: CaptureCapabilities,
        preferences: [String]
    ) -> Evaluation {
        guard acquisition.strategy == "any_of" || acquisition.strategy == "all_of" else {
            return Evaluation(methods: nil, unavailable: [])
        }
        let methods = acquisition.methods
        guard !methods.isEmpty, Set(methods).count == methods.count else {
            return Evaluation(methods: nil, unavailable: [])
        }
        let supported = methods.filter(capabilities.declaredMethods.contains)
        let available = methods.filter(capabilities.availableMethods.contains)
        let ordered = orderedMethods(available, preferences: preferences)
        if acquisition.strategy == "any_of", !ordered.isEmpty {
            return Evaluation(methods: ordered, unavailable: [])
        }
        if acquisition.strategy == "all_of", ordered.count == methods.count {
            return Evaluation(methods: ordered, unavailable: [])
        }
        var unavailable: [CaptureFallbackCondition] = []
        if supported.count != methods.count { unavailable.append(.methodUnavailable) }
        if supported.contains(where: { !capabilities.availableMethods.contains($0) }) {
            unavailable.append(.capabilityUnavailable)
        }
        return Evaluation(methods: nil, unavailable: unavailable)
    }

    private static func orderedMethods(_ methods: [String], preferences: [String]) -> [String] {
        guard !preferences.isEmpty else { return methods }
        let preferred = preferences.filter(methods.contains)
        guard !preferred.isEmpty else { return methods }
        return preferred + methods.filter { !preferred.contains($0) }
    }

    private static func fallbackCondition(
        _ fallback: CaptureFallback,
        unavailable: [CaptureFallbackCondition]
    ) -> CaptureFallbackCondition? {
        for value in fallback.on {
            guard let condition = CaptureFallbackCondition(rawValue: value), condition != .captureFailed else { continue }
            if unavailable.contains(condition) { return condition }
        }
        return nil
    }

    private static func makeTasks(
        requirement: CaptureRequirementDocument,
        methods: [String],
        fallbackCondition: CaptureFallbackCondition?,
        artefacts: [String]? = nil
    ) -> [CaptureTask] {
        let values = artefacts ?? requirement.artefacts
        if methods.count == 1 {
            return values.map { task(requirement, artefact: $0, methods: methods, fallback: fallbackCondition) }
        }
        let strategy = requirement.acquisition.strategy
        if strategy == "any_of" {
            return values.map { task(requirement, artefact: $0, methods: methods, fallback: fallbackCondition) }
        }
        return values.flatMap { artefact in
            methods.map { method in task(requirement, artefact: artefact, methods: [method], fallback: fallbackCondition) }
        }
    }

    private static func applyingCaptureFailure(
        requirement: CaptureRequirementDocument,
        tasks: [CaptureTask],
        capabilities: CaptureCapabilities,
        preferences: [String],
        failedTaskIDs: Set<String>
    ) throws -> [CaptureTask] {
        guard !failedTaskIDs.isEmpty, tasks.contains(where: { failedTaskIDs.contains($0.id) }) else { return tasks }
        let failedArtefacts = Set(tasks.filter { failedTaskIDs.contains($0.id) }.map(\.artefact))
        var replacement: [String: [CaptureTask]] = [:]
        for fallback in requirement.fallbacks where fallback.on.contains(CaptureFallbackCondition.captureFailed.rawValue) {
            let evaluation = evaluate(fallback.acquisition, capabilities: capabilities, preferences: preferences)
            guard let methods = evaluation.methods else { continue }
            for artefact in failedArtefacts where replacement[artefact] == nil {
                replacement[artefact] = makeTasks(
                    requirement: requirement,
                    methods: methods,
                    fallbackCondition: .captureFailed,
                    artefacts: [artefact]
                )
            }
        }
        guard failedArtefacts.allSatisfy({ replacement[$0] != nil }) else {
            throw IdenqaError.noCompatibleMethod
        }
        return tasks.flatMap { task in replacement[task.artefact] ?? [task] }
    }

    private static func task(
        _ requirement: CaptureRequirementDocument,
        artefact: String,
        methods: [String],
        fallback: CaptureFallbackCondition?
    ) -> CaptureTask {
        let id = taskID(
            requirementKey: requirement.key,
            artefact: artefact,
            methods: methods,
            fallback: fallback
        )
        return CaptureTask(
            id: id,
            requirementKey: requirement.key,
            evidenceType: requirement.evidenceType,
            artefact: artefact,
            methodOptions: methods,
            fallbackCondition: fallback,
            requiredAssurances: requirement.requiredAssurances
        )
    }

    static func taskID(
        requirementKey: String,
        artefact: String,
        methods: [String],
        fallback: CaptureFallbackCondition?
    ) -> String {
        [requirementKey, artefact, methods.joined(separator: "+"), fallback?.rawValue ?? ""]
            .joined(separator: "::")
    }

    private static func completionIDs(
        tasks: [CaptureTask],
        progress: [CaptureCompletionDocument]
    ) throws -> Set<String> {
        var completed: Set<String> = []
        for completion in progress {
            let matches = tasks.filter { task in
                task.requirementKey == completion.requirementKey &&
                    task.evidenceType == completion.evidenceType &&
                    task.artefact == completion.artefact &&
                    task.methodOptions.contains(completion.acquisitionMethod) &&
                    (task.fallbackCondition?.rawValue ?? "") == (completion.fallbackCondition ?? "")
            }
            guard matches.count == 1 else { throw IdenqaError.invalidResponse }
            guard completed.insert(matches[0].id).inserted else { throw IdenqaError.invalidResponse }
        }
        return completed
    }
}
